amd_openvx/openvx/ago/ago_haf_gpu_common.cpp

```diff
@@ -245,4 +245,227 @@ int HafGpu_Load_Local(int WGWidth, int WGHeight, int LMWidth, int LMHeight, int
+	if (dGroups == 1 && LMWidthRemain > 0 && LMHeightRemain > 0 && LMRemain < (WGWidth * WGHeight)) {
+		// sanity check
+		if (LMWidthRemain & ((1 << dTypeShift) - 1)) {
+			agoAddLogEntry(NULL, VX_FAILURE, "ERROR: HafGpu_Load_Local(%dx%d,%dx%d,(%d,%d)): doesn't support LMWidthRemain=%d with %s\n", WGWidth, WGHeight, LMWidth, LMHeight, gxoffset, gyoffset, LMWidthRemain, dType);
+			return -1;
+		}
+		if (use_vload) {
+			sprintf(item, "    *(__local %s *)(lbuf + loffset) = vload%c(0, (__global uint *)(gbuf + goffset));\n", dType, dType[4]);
+		}
+		else {
+			sprintf(item, "    *(__local %s *)(lbuf + loffset) = *(__global %s *)(gbuf + goffset);\n", dType, dType);
+		}
+		code += item;
+		// get configuration for extra load
+		int dWidth = LMWidthRemain >> dTypeShift;
+		int dHeight = LMHeight;
+		int dSize = dWidth * dHeight;
+		int dWidthShift = leftmostbit(dWidth);
+		if (dWidth != (1 << dWidthShift)) dWidthShift = -1;
+		sprintf(item,
+			OPENCL_FORMAT(
+			"    bool doExtraLoad = false;\n"
+			"    if (ly < %d) {\n" // LMHeight - WGHeight
+			"      loffset += %d * %d;\n" // WGHeight, LMWidth
+			"      goffset += %d * gstride;\n" // WGHeight
+			"      doExtraLoad = true;\n"
+			"    }\n"
+			"    else {\n"
+			"      int id = (ly - %d) * %d + lx;\n" // LMHeight - WGHeight, WGWidth
+			"      int ry = id %s %d;\n" // (id / dWidth) or (id >> dWidthShift)
+			"      int rx = id %s %d;\n" // (id - ry * dWidth) or (id & (dWidth-1))
+			"      loffset = ry * %d + (rx << %d) + %d;\n" // LMWidth, dTypeShift
+			"      goffset = (gy - ly + ry - %d) * gstride + ((gx - lx + rx) << %d) + %d;\n" // gyoffset, dTypeShift, (WGWidth << LMdivWGWidthShift) - gxoffset
+			"      doExtraLoad = (ry < %d) ? true : false;\n" // LMHeight
+			"    }\n"
+			"    if (doExtraLoad) {\n")
+			, LMHeight - WGHeight, WGHeight, LMWidth, WGHeight
+			, LMHeight - WGHeight, WGWidth, (dWidthShift < 0) ? "/" : ">>", (dWidthShift < 0) ? dWidth : dWidthShift, (dWidthShift < 0) ? "- ry *" : "&", (dWidthShift < 0) ? dWidth : dWidth - 1
+			, LMWidth, dTypeShift, (WGWidth << LMdivWGWidthShift)
+			, gyoffset, dTypeShift, (WGWidth << LMdivWGWidthShift) - gxoffset, LMHeight);
+		code += item;
+		if (use_vload) {
+			sprintf(item, "      *(__local %s *)(lbuf + loffset) = vload%c(0, (__global uint *)(gbuf + goffset));\n", dType, dType[4]);
+		}
+		else {
+			sprintf(item, "      *(__local %s *)(lbuf + loffset) = *(__global %s *)(gbuf + goffset);\n", dType, dType);
+		}
+		code += item;
+		code += "    }\n";
+	}
+	else {
+		for (int y = 0; y < LMHeight; y += WGHeight) {
+			if ((LMHeight - y) < WGHeight) {
+				sprintf(item, "    if (ly < %d) {\n", LMHeight - y);
+				code += item;
+			}
+			if (y > 0) {
+				sprintf(item,
+					"    loffset += %d * %d;\n" // WGHeight, LMWidth
+					"    goffset += %d * gstride;\n" // WGHeight
+                    "    int sidx = gbuf + goffset - p1_buf - p1_offset;\n"
```

Comment:
where is p1_buf and p1_offset calculated? If it is calculated before, please update comment in the beginning to explain those